subprojects/core/src/main/java/org/gradle/internal/work/DefaultAsyncWorkTracker.java

```diff
@@ -67,31 +67,51 @@ public void waitForCompletion(BuildOperationRef operation, boolean releaseLocks)
         }
 
         try {
-            if (workItems.size() > 0) {
-                boolean workInProgress = CollectionUtils.any(workItems, new Spec<AsyncWorkCompletion>() {
-                    @Override
-                    public boolean isSatisfiedBy(AsyncWorkCompletion workCompletion) {
-                        return !workCompletion.isComplete();
-                    }
-                });
-                // only release the project lock if we have to wait for items to finish
-                if (releaseLocks && workInProgress) {
-                    projectLeaseRegistry.withoutProjectLock(
-                        new Runnable() {
-                        @Override
-                        public void run() {
-                            waitForItemsAndGatherFailures(workItems);
-                        }
-                    });
-                } else {
-                    waitForItemsAndGatherFailures(workItems);
-                }
+            if (!workItems.isEmpty()) {
+                waitForItemsAndGatherFailures(workItems, lockRetention);
             }
         } finally {
             stopWaiting(operation);
         }
     }
 
+    private void waitForItemsAndGatherFailures(List<AsyncWorkCompletion> workItems, AsyncWorkTracker.ProjectLockRetention lockRetention) {
+        boolean workInProgress = hasWorkInProgress(workItems);
+
+        if (!workInProgress) {
```

Comment:
I'm happy to make that change, although I don't think it will have a practical impact: instead of releasing immediately we'd hold the lock until task execution is completed. But there's no more user code to invoke in this case, so I presume the difference will be small. BTW where is the code that releases the project lock after task execution is completed?

Reply:
Fixed in 878a0241d6b09701edc0bcac153a4aaf2daa7e2d

Reply:
It's here: https://github.com/gradle/gradle/blob/d35b309b56963f75f6c378c59d4a95e39a647ed7/subprojects/core/src/main/java/org/gradle/execution/plan/DefaultExecutionPlan.java#L887

Agreed that it will be a small difference (if any) in most cases.  It could have a larger effect, though, if snapshotting takes a long time, etc.   My biggest concern here is just consistent behavior for a given task.